### PR TITLE
Change backward_guard to optimize_guard to maximize the allreduce overlap.

### PR DESCRIPTION
**Pull request by gongweibao**
1. Change backward_guard to optimize_guard to maximize the allreduce overlap.
If we put `cast` operator in backward stage， it will wait for allredudce completion, and since paddle has only one compute stream, the stream will hang for waiting. So we need to put cast operator in optimize stage to avoid this.
image

2. Add more logs to locate potential problem.

### CODE REVIEW

**Review by guru4elephant — COMMENTED** [review]
I think you can explain the reason why this pr can maximize the allreduce overlap with some figures or examples.

**Comment by gongweibao** [comment]
Done.Thanks!

**Review by kuke — COMMENTED** [review]
LGTM for the optimize_guard

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py [hunk 1]
```diff
@@ -80,8 +80,9 @@ def create_master_params_grads(params_grads, main_prog, startup_prog,
         A list of master parameters and gradients. 
     """
     master_params_grads = []
-    with main_prog._backward_role_guard():
-        for p, g in params_grads:
+    for p, g in params_grads:
+        # create master parameters
+        with main_prog._optimized_guard([p, g]):
```
Comment (gavin1332) [inline python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:85]: what's the difference while training with pe? As the op orders in the front-end have no effect on the dependency.

Comment (gongweibao) [reply python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:85]: It doesn't change the dependency, but delay op's running to avoid waiting.

**Review by gavin1332 — APPROVED** [review]
LGTM

**Pull request merged**